### ISSUE TITLE
Setting `shuffle=True` for `val_dataloader` doesn't work when `ddp` enabled.

### ISSUE THREAD

**Issue by cjwcommuny** [PyTorchLightning/pytorch-lightning#11054]
## 🐛 Bug

Setting `shuffle=True` for `val_dataloader` doesn't work when `ddp` enabled.

### To Reproduce

minimum snippet:

```python
from pytorch_lightning import LightningModule, Trainer
from torch import nn
from torch.optim import Adam
from torch.utils.data import DataLoader


class PlModule(LightningModule):
    def __init__(self):
        super().__init__()
        self.model = nn.Linear(1, 1)
        self.train_dataset = list(range(1000))
        self.val_dataset = list(range(1000))

    def configure_optimizers(self):
        return Adam(self.model.parameters(), lr=0.1)

    def training_step(self, batch, *args, **kwargs):
        return self.model(batch)

    def validation_step(self, batch, *args, **kwargs):
        print(f'{batch=}')
        print(f'{self.trainer.val_dataloaders[0].sampler.shuffle=}')
        exit(0)

    def train_dataloader(self):
        return DataLoader(self.train_dataset, shuffle=True, batch_size=5)

    def val_dataloader(self):
        return DataLoader(self.val_dataset, shuffle=True, batch_size=5)



if __name__ == '__main__':
    module = PlModule()
    trainer = Trainer(
        limit_train_batches=0.0,
        limit_val_batches=1.0,
        num_sanity_val_steps=-1,
        gpus=2,
        strategy='ddp',
        enable_model_summary=False
    )
    trainer.fit(module)
```

By execute the snippet above, we get the stdout:

```
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/2
initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 2 processes
----------------------------------------------------------------------------------------------------

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1,2,3,4,5,6,7]
Validation sanity check: 0it [00:00, ?it/s]/home/balabala/miniconda3/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:452: UserWarning: Your `val_dataloader` has `shuffle=True`,it is strongly recommended that you turn this off for val/test/predict dataloaders.
  rank_zero_warn(
/home/balabala/miniconda3/lib/python3.8/site-packages/pytorch_lightning/trainer/data_loading.py:110: UserWarning: The dataloader, val_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 96 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Validation sanity check:   0%|                          | 0/100 [00:00<?, ?it/s]batch=tensor([1, 3, 5, 7, 9], device='cuda:1')
self.trainer.val_dataloaders[0].sampler.shuffle=False
batch=tensor([0, 2, 4, 6, 8], device='cuda:0')
self.trainer.val_dataloaders[0].sampler.shuffle=False
```

### Expected behavior

the stdout should contains the follow:

```
batch= # some random batch
self.trainer.val_dataloaders[0].sampler.shuffle=True
```

### Environment

- PyTorch Lightning Version 1.5.2
- PyTorch Version 1.10.0
- Python version 3.8.10

cc @borda @tchaton @rohitgr7 @justusschock @awaelchli @ninginthecloud @kaushikb11 @akihironitta

**Comment by ananthsub** [comment]
Out of curiosity, could you describe why you need shuffling during validation?

**Comment by cjwcommuny** [comment]
> Out of curiosity, could you describe why you need shuffling during validation?

In training, I would like to construct random positive-negative pairs on the fly. In validation, I want to keep the same setting. Of course, I can construct random data before model fitting and fix it, but it's more convenient to do it on the fly, and I don't have to write additional code. Since I would like to choose a large batch size, the randomness the validation setting brings is acceptable.

**Comment by bradyz** [comment]
In your model's `pl.LightningModule`,

```python
class ModelModule(pl.LightningModule):
    def on_validation_start(self):
        for v in self.trainer.val_dataloaders:
            v.sampler.set_epoch(self.current_epoch)
```

**Comment by ananthsub** [comment]
This is the line in the Trainer where shuffling is hardcoded to be False for validation, test, and prediction dataloaders: https://github.com/PyTorchLightning/pytorch-lightning/blob/1c5a5c3dfeeefa7b8760ab5f5869f3f61c48e665/pytorch_lightning/trainer/data_loading.py#L312-L313

**Comment by tchaton** [comment]
@rohitgr7 Mind having a look.

**Comment by bsridatta** [comment]
@ananthsub I was looking into similar issue earlier and before I made a PR the structure was changed. 
The logic is now in `DataConnector` and the changes I made previously are no longer enough to shuffle. 

The self._check_eval_shuffling function below doesn't return anything other than raising an error.
https://github.com/PyTorchLightning/pytorch-lightning/blob/9cf9ded73b66296e1b22369ea4e8327f9ce0aae8/pytorch_lightning/trainer/connectors/data_connector.py#L429-L435

https://github.com/PyTorchLightning/pytorch-lightning/blob/9cf9ded73b66296e1b22369ea4e8327f9ce0aae8/pytorch_lightning/trainer/connectors/data_connector.py#L532-L543

As mentioned in the trainer setting `replace_sampler_ddp` might help in a few cases.
https://github.com/PyTorchLightning/pytorch-lightning/blob/16a04b29eb811f105a11db4f7f3003489252f9f8/pytorch_lightning/trainer/trainer.py#L361-L364
